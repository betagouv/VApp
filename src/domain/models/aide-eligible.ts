import { assertValid } from '@/domain/note';

export class AideEligible {
  static SELECTION = 10;

  constructor(
    public eligibilite: number,
    public aideId: string
  ) {
    assertValid(eligibilite);

    this.eligibilite = eligibilite;
    this.aideId = aideId;
  }

  static compare(a: Pick<AideEligible, 'eligibilite'>, b: Pick<AideEligible, 'eligibilite'>) {
    return b.eligibilite - a.eligibilite;
  }
}
