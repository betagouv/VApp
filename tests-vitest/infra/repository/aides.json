[
  { "uuid": "v4ALdi6YbmK4ng9yw2C28i", "atId": "aide-a", "nom": "Aide A", "description": "" },
  { "uuid": "rMuvuLcZ4XmedH2bgFjvY2", "atId": "aide-b", "nom": "Aide B", "description": "" },
  { "uuid": "2XekLaps12ztaiAiVC3FzM", "atId": "aide-C", "nom": "Aide C", "description": "" },
  { "uuid": "bVC26eSfGYdBpF9zsCMC5D", "atId": "aide-D", "nom": "Aide D", "description": "" }
]
