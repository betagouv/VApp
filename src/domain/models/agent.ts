import { Projet } from './projet';

export class Agent {
  public uuid: string;
  public projects: Projet[] = [];

  constructor(uuid: string, projects: Projet[] = []) {
    this.uuid = uuid;
    this.projects = projects;
  }
}
