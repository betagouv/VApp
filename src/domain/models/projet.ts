import short, { SUUID, UUID } from 'short-uuid';
import { AideScore } from '@/domain/models/aide-score';
import { Porteur } from '@/domain/models/projet/porteur';
import { LesCommunsProjetStatuts } from '@/domain/models/les-communs/projet-statuts';
import { ZoneGeographique } from '@/domain/models/zone-geographique';
import { AideScoreMap } from '@/domain/models/aide-score.map';
import { CriteresRechercheAide } from '@/domain/models/criteres-recherche-aide';

const translator = short();

export class Projet {
  public criteresRechercheAide: CriteresRechercheAide = {};

  constructor(
    public uuid: UUID,
    public description: string,
    public porteur: Porteur,
    public etatAvancement: LesCommunsProjetStatuts = LesCommunsProjetStatuts.IDEE,
    public zonesGeographiques: ZoneGeographique[] = [],
    public aidesScores: AideScoreMap = new Map(),
    public clientId?: UUID
  ) {}

  public static create(
    description: string,
    porteur: Porteur,
    etatAvancement: LesCommunsProjetStatuts = LesCommunsProjetStatuts.IDEE,
    zonesGeographiques: ZoneGeographique[] = [],
    aidesScores: AideScoreMap = new Map(),
    clientId?: UUID,
    uuid?: UUID
  ): Projet {
    if (!description || description.length === 0) {
      throw new Error('La description du projet est vide.');
    }

    return new Projet(
      uuid || translator.uuid(),
      description,
      porteur,
      etatAvancement,
      zonesGeographiques,
      aidesScores,
      clientId
    );
  }

  public get suuid(): SUUID {
    return translator.fromUUID(this.uuid);
  }

  public reformuler(description: Projet['description']) {
    this.description = description;
    this.aidesScores = new Map();
  }

  public updateScores(scores: AideScore[]): void {
    scores.forEach((score) => {
      this.aidesScores.set(score.aideId, score);
    });
  }

  public addScore(score: AideScore): void {
    this.aidesScores.set(score.aideId, score);
  }

  public getSortedAideScores() {
    return Array.from(this.aidesScores, ([, aideScore]) => aideScore).sort(AideScore.compare);
  }
}
