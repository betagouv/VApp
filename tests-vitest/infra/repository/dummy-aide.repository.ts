import { AideRepositoryInterface } from '../../../src/domain/repositories/aide.repository.interface';
import { Aide } from '../../../src/domain/models/aide';
import aides from './aides.json';

export class DummyAideRepository implements AideRepositoryInterface {
  constructor(public aides: Aide[] = []) {}

  public all() {
    return Promise.resolve(this.aides);
  }

  public async fromUuid(uuid: string): Promise<Aide> {
    const aides = await this.all();
    const aide = aides.find((aide) => aide.uuid === uuid);
    if (!aide) {
      throw new Error(`Aucunes aides ne portent l'identifiant ${uuid}.`);
    }

    return aide;
  }
}

// @ts-ignore
export const dummyAideRepository = new DummyAideRepository(aides);
