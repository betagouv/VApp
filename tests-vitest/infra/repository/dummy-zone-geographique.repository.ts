import { ZoneGeographiqueRepositoryInterface } from '@/domain/repositories/zone-geographique-repository.interface';
import { ZoneGeographique } from '@/domain/models/zone-geographique';
import { AtPerimeterScale } from '@/infra/at/perimeter';

export class DummyZoneGeographiqueRepository implements ZoneGeographiqueRepositoryInterface {
  all(): Promise<ZoneGeographique[]> {
    return Promise.resolve([]);
  }

  autocomplete(query: string): Promise<ZoneGeographique[]> {
    return Promise.resolve([]);
  }

  findOneByTypeAndCode(type: AtPerimeterScale, code: string): Promise<ZoneGeographique | null> {
    return Promise.resolve(null);
  }

  fromId(atId: ZoneGeographique['id']): Promise<ZoneGeographique> {
    return Promise.resolve({
      id: atId,
      nom: 'test',
      description: 'string',
      code: 'string',
      type: AtPerimeterScale.commune
    });
  }

  size(): Promise<number> {
    return Promise.resolve(0);
  }
}

// @ts-ignore
export const dummyZoneGeographiqueRepository = new DummyZoneGeographiqueRepository();
