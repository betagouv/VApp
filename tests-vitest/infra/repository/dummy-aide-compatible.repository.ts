import { PaginatedResult, Pagination } from '@/libs/pagination';
import { Aide } from '@/domain/models/aide';
import { Projet } from '@/domain/models/projet';
import { AideScore } from '@/domain/models/aide-score';
import { CriteresRechercheAide } from '@/domain/models/criteres-recherche-aide';
import { AideScoreRepositoryInterface } from '@/domain/repositories/aide-score-repository.interface';
import aides from './aides.json';

export class DummyAideCompatibleRepository implements AideScoreRepositoryInterface {
  constructor(public aides: Aide[] = []) {}

  public all() {
    return Promise.resolve(this.aides.map(({ id }, i) => new AideScore(i, id)));
  }

  public size() {
    return Promise.resolve(this.aides.length);
  }

  public findAllFor() {
    return this.all();
  }

  async *findAllAsyncForProjet(
    projet: Projet,
    criteres: CriteresRechercheAide,
    chunkSize?: number
  ): AsyncGenerator<AideScore> {
    return undefined;
  }

  async findAllPaginatedForProjet(projet: Projet, { page, limit }: Pagination): Promise<PaginatedResult<AideScore>> {
    const aidesCompatibles = await this.all();
    return Promise.resolve({
      data: aidesCompatibles,
      totalItems: aidesCompatibles.length,
      lastPage: 1,
      currentPage: page || 1
    });
  }
}

// @ts-ignore
export const dummyAideCompatibleRepository = new DummyAideCompatibleRepository(aides);
