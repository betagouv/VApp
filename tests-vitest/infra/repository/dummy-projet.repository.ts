import { ProjetRepositoryInterface } from '@/domain/repositories/projet.repository.interface';
import { Projet } from '@/domain/models/projet';

export class DummyProjetRepository implements ProjetRepositoryInterface {
  constructor(public projets: Projet[] = []) {}

  public all() {
    return Promise.resolve(this.projets);
  }

  public async fromUuid(uuid: string): Promise<Projet> {
    const projets = await this.all();
    const projet = projets.find((projet) => projet.uuid === uuid);
    if (!projet) {
      throw new Error(`Aucuns projets ne portent l'identifiant ${uuid}.`);
    }

    return projet;
  }

  add(projet: Projet): Promise<void> {
    this.projets = this.projets.concat([projet]);

    return Promise.resolve();
  }
}

// @ts-ignore
export const dummyProjetRepository = new DummyProjetRepository();
