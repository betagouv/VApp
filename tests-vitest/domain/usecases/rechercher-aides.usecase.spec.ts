import { it, describe, expect } from 'vitest';
import { RechercherAidesUsecase } from '../../../src/domain/usecases/rechercher-aides.usecase';
import { dummyAideRepository } from '../../infra/repository/dummy-aide.repository';
import { randomNotationAideService } from '../../infra/services/random-notation-aide.service';
import { Projet } from '../../../src/domain/models/projet';

describe('suggerer aide usecase', () => {
  it('throws when the project description is empty', async () => {
    const projet = Projet.create('');
    const suggererAides = new RechercherAidesUsecase(randomNotationAideService, dummyAideRepository);
    await expect(suggererAides.execute(projet)).rejects.toThrowError(/description/);
  });

  it('returns a list of suggestions', async () => {
    const projet = Projet.create('Restaurer une zone humide');
    const suggererAides = new RechercherAidesUsecase(randomNotationAideService, dummyAideRepository);
    const suggestions = await suggererAides.execute(projet);
    suggestions.forEach((suggestion) =>
      expect(suggestion).toMatchSnapshot({
        eligibilite: expect.any(Number),
        aideId: expect.any(String)
      })
    );
  });
});
