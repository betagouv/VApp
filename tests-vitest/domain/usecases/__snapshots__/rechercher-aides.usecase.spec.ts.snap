// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`poser questions usecase > returns a list of suggestions 1`] = `
{
  "aideId": Any<String>,
  "eligibilite": Any<Number>,
}
`;

exports[`poser questions usecase > returns a list of suggestions 2`] = `
{
  "aideId": Any<String>,
  "eligibilite": Any<Number>,
}
`;

exports[`poser questions usecase > returns a list of suggestions 3`] = `
{
  "aideId": Any<String>,
  "eligibilite": Any<Number>,
}
`;
